FWCore/Framework/src/PathsAndConsumesOfModules.cc

```diff
@@ -288,196 +352,345 @@ namespace edm {
+    std::vector<unsigned int> stackTrace;
+    bool madeForwardProgress = true;
+    try {
+      //'simulate' the running of the paths. On each step mark each module as 'run'
+      // if all the module's dependencies were fulfilled in a previous step
+      unsigned long long searchIndex = 0;
+      while (madeForwardProgress and nPathsFinished != statusOfPaths.size()) {
+        madeForwardProgress = false;
+        for (auto& p : statusOfPaths) {
+          //the path has already completed in an earlier pass
+          if (p.activeModuleSlot_ == p.nModules_) {
+            continue;
+          }
+          ++searchIndex;
+          bool didRun = checkIfCanRun(searchIndex, p.modulesOnPath_[p.activeModuleSlot_], statusOfModules, stackTrace);
+          if (didRun) {
+            madeForwardProgress = true;
+            ++p.activeModuleSlot_;
+            if (p.activeModuleSlot_ == p.nModules_) {
+              ++nPathsFinished;
+            }
+          }
+        }
+      }
+    } catch (CircularDependencyException const&) {
+      //the last element in stackTrace must appear somewhere earlier in stackTrace
+      std::ostringstream oStr;
+
+      unsigned int lastIndex = stackTrace.front();
+      bool firstSkipped = false;
+      for (auto id : stackTrace) {
+        if (firstSkipped) {
+          oStr << "  module '" << moduleIndexToNames[lastIndex] << "' depends on " << moduleIndexToNames[id] << "\n";
+        } else {
+          firstSkipped = true;
+        }
+        lastIndex = id;
+      }
+      throw edm::Exception(edm::errors::ScheduleExecutionFailure, "Unrunnable schedule\n")
+          << "Circular module dependency found in configuration\n"
+          << oStr.str();
+    }
 
-    //determine the path dependencies
-    std::vector<std::string> pathNames = iPnC.paths();
-    const unsigned int kFirstEndPathIndex = pathNames.size();
+    auto pathName = [&](PathStatus const& iP) {
+      if (iP.endPath_) {
+        return iPnC.endPaths()[iP.index_];
+      }
+      return iPnC.paths()[iP.index_];
+    };
 
-    const std::string kPathEnded("@PathEnded");
-    const std::string kEndPathStart("@EndPathStart");
+    //The program would deadlock
+    if (not madeForwardProgress) {
+      std::ostringstream oStr;
+      auto modIndex = std::numeric_limits<unsigned int>::max();
+      unsigned int presentPath;
+      for (auto itP = statusOfPaths.begin(); itP != statusOfPaths.end(); ++itP) {
+        auto const& p = *itP;
+        if (p.activeModuleSlot_ == p.nModules_) {
+          continue;
+        }
+        //this path is stuck
+        modIndex = p.modulesOnPath_[p.activeModuleSlot_];
+        presentPath = itP - statusOfPaths.begin();
+        break;
+      }
+      //NOTE the following should always be true as at least 1 path should be stuc.
+      // I've added the condition just to be paranoid.
+      if (modIndex != std::numeric_limits<unsigned int>::max()) {
```

Comment:
Usually I don't like comments at braces that end blocks of code, but it might be worth adding one at the end of this block and maybe a few of the other very long blocks. It took me a while to find the end of it.  Or better, this big function might be broken into a few smaller functions... I didn't get to the end of the PR yet. I'll start on this again tomorrow morning... In general, this looks like an improvement. It's easier to understand, and looks like it ought to be faster. I suppose testing is the only way to prove it.